security.rst

```diff
@@ -662,43 +360,44 @@ The process of authorization has two different sides:
    "attribute" (most commonly a role like ``ROLE_ADMIN``) in order to be
    accessed.
 
-.. tip::
-
-    In addition to roles (e.g. ``ROLE_ADMIN``), you can protect a resource
-    using other attributes/strings (e.g. ``EDIT``) and use voters to give these
-    meaning. This might come in handy if you need to check if user A can "EDIT"
-    some object B (e.g. a Product with id 5). See :ref:`security-secure-objects`.
-
 Roles
 ~~~~~
 
-When a user logs in, they receive a set of roles (e.g. ``ROLE_ADMIN``). In
-the example above, these are hardcoded into ``security.yaml``. If you're
-loading users from the database, these are probably stored on a column
-in your table.
+When a user logs in, Symfony calls the ``getRoles()`` method on your ``User``
+object to determime which roles this user has. In the ``User`` class that we
+generated earlier, the roles are an array that's stored in the database, and
+every user is *always* given at least one role: ``ROLE_USER``::
+
+    // src/Entity/User.php
+    // ...
+
+    /**
+     * @ORM\Column(type="json")
+     */
+    private $roles = [];
+
+    public function getRoles(): array
+    {
+        $roles = $this->roles;
+        // guarantee every user at least has ROLE_USER
+        $roles[] = 'ROLE_USER';
 
-.. caution::
+        return array_unique($roles);
+    }
 
-    All roles you assign to a user **must** begin with the ``ROLE_`` prefix.
-    Otherwise, they won't be handled by Symfony's security system in the
-    normal way (i.e. unless you're doing something advanced, assigning a
-    role like ``FOO`` to a user and then checking for ``FOO`` as described
-    :ref:`below <security-role-authorization>` will not work).
+This is a nice default, but you can do *whatever* you want to determine which roles
+a user should have. Here are a few guidelines:
 
-Roles are simple, and are basically strings that you invent and use as needed.
-For example, if you need to start limiting access to the blog admin section
-of your website, you could protect that section using a ``ROLE_BLOG_ADMIN``
-role. This role doesn't need to be defined anywhere - you can just start using
-it.
+* Every role **must start with** ``ROLE_`` (otherwise, things won't as expected)
 
-.. tip::
+* Other than the above rule, a role is just a string and you can invent what you
+  need (e.g. ``ROLE_PRODUCT_ADMIN``)
 
-    Make sure every user has at least *one* role, or your user will look
-    like they're not authenticated. A common convention is to give *every*
-    user ``ROLE_USER``.
+* Every User **must** have at least **one** role - a common convention is to give
```

Comment:
Ah! I didn't know that Guard fixed the "one role" problem - that's awesome! :D

> In fact, I think the docs should encourage using IS_AUTHENTICATED_ANONYMOUSLY, IS_AUTHENTICATED_REMEMBERED and IS_AUTHENTICATED_FULLY for access control instead of things like ROLE_USER.

This is something we should definitely discuss. I would be a bit simpler, but also, wow, `IS_AUTHENTICATED_FULLY` is WAY too long. If we could add a shorter alias in core, I would LOVE this idea.

> I would also like to see a mention that roles are an extension to identification if possible and it's not intended to directly grant permissions via roles like ROLE_EDIT_POST.

I'm not sure the best way to say this. I also recommend this approach: using role names that more describe *who* you are (e.g. `ROLE_PRODUCT_MANAGER` or `ROLE_USER_ADMIN`) rather than exactly what you can do (`ROLE_EDIT_PRODUCT`). But that requires leveraging role_hierarchy. So, to fully explain this, that feature might need to be moved up earlier... which is unfortunate.